internal/dag/extension_processor.go

```diff
@@ -0,0 +1,132 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dag
+
+import (
+	"path"
+
+	"github.com/projectcontour/contour/apis/projectcontour/v1alpha1"
+	"github.com/projectcontour/contour/internal/k8s"
+	"github.com/projectcontour/contour/internal/xds"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+type ExtensionServiceProcessor struct {
+	logrus.FieldLogger
+}
+
+var _ Processor = &ExtensionServiceProcessor{}
+
+func (p *ExtensionServiceProcessor) Run(builder *Builder) {
+	for n, e := range builder.Source.extensions {
+		if ext := p.buildExtensionService(builder, e); ext != nil {
+			builder.extensions[n] = ext
+		}
+	}
+}
+
+// buildExtensionService builds one ExtensionService record based
+// on the corresponding CRD.
+//
+// TODO: Publish status conditions in https://github.com/projectcontour/contour/issues/2875.
```

Comment:
Nit: Can you please tag `TODO` items with your github name? So `TODO(jpeach)` - that means that if anyone wants to pick it up, they should check with you about the context.